ntp-daemon/src/spawn/mod.rs

```diff
@@ -0,0 +1,240 @@
+use std::{net::SocketAddr, sync::atomic::AtomicU64};
+
+use ntp_proto::PeerNtsData;
+use tokio::sync::mpsc;
+
+use crate::config::NormalizedAddress;
+
+#[cfg(test)]
+pub mod dummy;
+pub mod nts;
+pub mod pool;
+pub mod standard;
+
+/// Unique identifier for a spawner
+/// This is used to identify which spawner was used to create a peer
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct SpawnerId(u64);
+
+impl SpawnerId {
+    pub fn new() -> SpawnerId {
+        static COUNTER: AtomicU64 = AtomicU64::new(1);
+        SpawnerId(COUNTER.fetch_add(1, std::sync::atomic::Ordering::Relaxed))
+    }
+}
+
+impl Default for SpawnerId {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Unique identifier for a peer created by a spawner
+/// This peer id makes sure that even if the network address is the same
+/// that we always know which specific spawned peer we are talking about.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct PeerId(u64);
+
+impl PeerId {
+    pub fn new() -> PeerId {
+        static COUNTER: AtomicU64 = AtomicU64::new(1);
+        PeerId(COUNTER.fetch_add(1, std::sync::atomic::Ordering::Relaxed))
+    }
+}
+
+impl Default for PeerId {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct SpawnEvent {
+    pub id: SpawnerId,
+    pub action: SpawnAction,
+}
+
+#[derive(Debug)]
+pub enum SystemEvent {
+    PeerRemoved(PeerRemovedEvent),
+    PeerRegistered(PeerCreateParameters),
+    Shutdown,
+}
+
+impl SystemEvent {
+    pub fn peer_removed(id: PeerId, reason: PeerRemovalReason) -> SystemEvent {
+        SystemEvent::PeerRemoved(PeerRemovedEvent { id, reason })
+    }
+
+    pub fn peer_registered(
+        id: PeerId,
+        addr: SocketAddr,
+        normalized_addr: NormalizedAddress,
+        nts: Option<PeerNtsData>,
+    ) -> SystemEvent {
+        SystemEvent::PeerRegistered(PeerCreateParameters {
+            id,
+            addr,
+            normalized_addr,
+            nts,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct PeerRemovedEvent {
+    pub id: PeerId,
+    pub reason: PeerRemovalReason,
+}
+
+#[derive(Debug)]
+pub enum PeerRemovalReason {
+    Demobilized,
+    NetworkIssue,
+}
+
+impl SpawnEvent {
+    pub fn new(id: SpawnerId, action: SpawnAction) -> SpawnEvent {
+        SpawnEvent { id, action }
+    }
+}
+
+/// The kind of action that the spawner requests to the system
+/// Currently a spawner can only create peers
+#[derive(Debug)]
+pub enum SpawnAction {
+    Create(PeerCreateParameters),
+    // Remove(()),
+}
+
+impl SpawnAction {
+    pub fn create(
+        id: PeerId,
+        addr: SocketAddr,
+        normalized_addr: NormalizedAddress,
+        nts: Option<PeerNtsData>,
+    ) -> SpawnAction {
+        SpawnAction::Create(PeerCreateParameters {
+            id,
+            addr,
+            normalized_addr,
+            nts,
+        })
+    }
+}
```

Comment:
this function seems kind of pointless?

Reply:
I noticed that having the enum and struct construction in the spawning algorithm made that little bit less readable than when this function was in place, especially given that creating the enum variant is more administration than anything algorithmically relevant. It would be my preference to keep it if you're ok with it.